src/Symfony/Component/HttpClient/ApiClient.php

```diff
@@ -0,0 +1,136 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpClient;
+
+use Symfony\Component\HttpClient\Response\ApiResponse;
+use Symfony\Component\HttpClient\Response\ApiResponseStream;
+use Symfony\Contracts\HttpClient\ApiClientInterface;
+use Symfony\Contracts\HttpClient\ApiResponseInterface;
+use Symfony\Contracts\HttpClient\ApiResponseStreamInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+
+/**
+ * @see ApiClientInterface::OPTIONS_DEFAULTS for available options
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+final class ApiClient implements ApiClientInterface
+{
+    use HttpClientTrait;
+
+    private $client;
+    private $defaultOptions = [
+        'headers' => [
+            'accept' => ['application/json'],
+        ],
+    ] + ApiClientInterface::OPTIONS_DEFAULTS;
+
+    /**
+     * A factory to instantiate the best possible API client for the runtime.
+     *
+     * @param array $defaultOptions     Default requests' options
+     * @param int   $maxHostConnections The maximum number of connections to a single host
+     *
+     * @see HttpClientInterface::OPTIONS_DEFAULTS for available options
+     */
+    public static function create(array $defaultOptions = [], int $maxHostConnections = 6): ApiClientInterface
+    {
+        return new self(HttpClient::create($defaultOptions, $maxHostConnections));
+    }
+
+    public function __construct(HttpClientInterface $client, array $defaultOptions = [])
+    {
+        $this->client = $client;
+
+        if ($defaultOptions) {
+            [, $this->defaultOptions] = self::prepareRequest(null, null, $defaultOptions, $this->defaultOptions);
```

Comment:
would it be possible to refactor the HttpClientTrait to have separate methods for the url and the options? i see that one section of the trait prepareRequest uses null === $url as a flag to do auth or not. i think this could be refactored too, to be more explicit and to have an options resolution that is not preparing the request.

Reply:
That's not completely possible as there is some coupling between options and url - the most obvious example is the `base_uri` options. I'd be happy to review a PR if you want to work on one.